### PR TITLE
Drop PyTorch 1.7 testing from the CI

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

For preparing the upcoming release, we plan to drop PyTorch 1.7 testing from the CI. Partially fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/12161.

### Does your PR introduce any breaking changes? If yes, please list them.

Waiting for the CI to turn green.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @akihironitta @carmocca @borda

### CODE REVIEW

File: .github/workflows/ci_test-full.yml [hunk 1]
```diff
@@ -24,9 +24,10 @@ jobs:
         python-version: ["3.7", "3.9"]  # minimum, maximum
         requires: ["oldest", "latest"]
         release: ["stable"]
-        exclude:
-          # Skip if torch<1.8 and py3.9 on Linux: https://github.com/pytorch/pytorch/issues/50014
-          - {os: ubuntu-20.04, python-version: "3.9", requires: "oldest"}
+        # exclude:
+          # TODO: @krshrimali, remove these comments if the CI is green now! (the linked issue is fixed and closed)
+          # Skip if py3.9 on Linux: https://github.com/pytorch/pytorch/issues/50014
+          # - {os: ubuntu-20.04, python-version: "3.9", requires: "oldest"}
```
Comment (carmocca) [inline .github/workflows/ci_test-full.yml:30]: You can remove these already

```suggestion
```

File: .github/workflows/ci_dockers.yml [hunk 2]
```diff
@@ -122,11 +122,9 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        # the config used in 'dockers/ipu-ci-runner/Dockerfile'
```
Comment (carmocca) [inline .github/workflows/ci_dockers.yml:125]: Mind keeping this comment?

File: .github/workflows/ci_dockers.yml [hunk 3]
```diff
@@ -122,11 +122,9 @@ jobs:
...
-        python_version: ["3.9"]  # latest
-        # TODO: upgrade - PopTorch 2.2 uses torch 1.9, see:
```
Comment (carmocca) [inline .github/workflows/ci_dockers.yml:127]: @Borda IPU CI installs PopTorch 2.1 which requires PyTorch 1.7

So for us to drop PyTorch 1.7, we need to use PyTorch 1.9 in IPU CI as Kush already did.

But where do we configure IPU CI to install PopTorch 2.2?

File: .github/workflows/ci_dockers.yml [hunk 4]
```diff
@@ -122,11 +122,9 @@ jobs:
...
-        # https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/installation.html#version-compatibility
-        pytorch_version: ["1.7"]
+        include:
+          - {python_version: "3.9", pytorch_version: "1.9"}
```
Comment (krshrimali) [inline .github/workflows/ci_dockers.yml:126]:
```suggestion
          # the config used in 'dockers/ipu-ci-runner/Dockerfile'
          - {python_version: "3.9", pytorch_version: "1.9"}
```

File: .github/workflows/ci_dockers.yml (hunk 2 above)
Comment (krshrimali) [reply .github/workflows/ci_dockers.yml:125]: Oops, sorry! Fixed now.

**Review by akihironitta — COMMENTED** [review]
The CI section in README.md needs updating, too.

File: .github/workflows/ci_test-conda.yml [hunk 5]
```diff
@@ -19,7 +19,7 @@ jobs:
       fail-fast: false
       matrix:
         python-version: ["3.8"]  # previous to last Python version as that one is already used in test-full
-        pytorch-version: ["1.7", "1.8", "1.9", "1.10"]  # nightly: add when there's a release candidate
+        pytorch-version: ["1.8", "1.9", "1.10"]  # nightly: add when there's a release candidate
```
Comment (krshrimali) [inline .github/workflows/ci_test-conda.yml:22]:
```suggestion
        pytorch-version: ["1.7", "1.8", "1.9", "1.10"]  # nightly: add when there's a release candidate
```

Rolling back some commits to see what went wrong.

File: .github/workflows/ci_dockers.yml [hunk 6]
```diff
@@ -97,7 +97,7 @@ jobs:
       matrix:
         # the config used in '.github/workflows/ci_test-conda.yml'
         python_version: ["3.8"]
-        pytorch_version: ["1.7", "1.8", "1.9", "1.10"]
+        pytorch_version: ["1.8", "1.9", "1.10"]
```
Comment (krshrimali) [inline .github/workflows/ci_dockers.yml:100]:
```suggestion
        pytorch_version: ["1.7", "1.8", "1.9", "1.10"]
```

Rolling back some commits to see what went wrong. https://github.com/PyTorchLightning/pytorch-lightning/runs/5411681687?check_suite_focus=true failing for 1.8, 1.9

**Review by krshrimali — COMMENTED** [review]
Horovod installation is failing. Debugging what went wrong.

File: .github/workflows/ci_dockers.yml [hunk 7]
```diff
@@ -124,9 +124,8 @@ jobs:
       matrix:
         # the config used in 'dockers/ipu-ci-runner/Dockerfile'
         python_version: ["3.9"]  # latest
-        # TODO: upgrade - PopTorch 2.2 uses torch 1.9, see:
-        # https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/installation.html#version-compatibility
-        pytorch_version: ["1.7"]
+        pytorch_version: ["1.9"]
+
```
Comment (krshrimali) [inline .github/workflows/ci_dockers.yml:128]:
```suggestion

```

**Comment by krshrimali** [comment]
> The CI section in README.md needs updating, too.

Thanks, @akihironitta! Done. https://github.com/PyTorchLightning/pytorch-lightning/blob/drop/pytorch-1.7/README.md

**Review by awaelchli — APPROVED** [review]
subsequently, all markers `@RunIf(min_torch="1.8")` or below can be dropped from our tests, here or in follow up. Tests with `@RunIf(max_torch="1.7")` can be dropped entriely.

You will find more references that end up in dead code or outdated comments if you grep torch 1.7

File: CHANGELOG.md [hunk 8]
```diff
@@ -146,6 +146,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
+- Drop PyTorch 1.7 testing from the CI ([#12191](https://github.com/PyTorchLightning/pytorch-lightning/pull/12191))
```
Comment (akihironitta) [inline CHANGELOG.md:149]:
```suggestion
- Drop PyTorch 1.7 support ([#12191](https://github.com/PyTorchLightning/pytorch-lightning/pull/12191))
```

File: environment.yml [hunk 9]
```diff
@@ -44,7 +44,7 @@ dependencies:
     - torchtext>=0.8.*
```
Comment (akihironitta) [inline environment.yml:44]: Should we update the min version of torchtext as well?
```suggestion
    - torchtext>=0.9.*
```

Same applies to:
https://github.com/PyTorchLightning/pytorch-lightning/blob/aea0a9b218aa9fe596684953eb8134154cf6d334/requirements/extra.txt#L5

(Its version matrix is here: https://github.com/pytorch/text#installation)

**Review by akihironitta — APPROVED** [review]

File: environment.yml (hunk 9 above)
Comment (carmocca) [reply environment.yml:44]: Yes

**Comment by carmocca** [comment]
@awaelchli thanks! We plan to do those in a separate PR as per the linked issue description.

File: environment.yml (hunk 9 above)
Comment (krshrimali) [reply environment.yml:44]: Thanks, @akihironitta - I missed this, it makes sense. :)

**Review by Borda — COMMENTED** [review]
I think first we need to add PT 1.11 #11792

**Review by Borda — APPROVED** [review]

**Pull request merged**